ipykernel/kernelbase.py

```diff
@@ -379,21 +304,88 @@ def should_handle(self, stream, msg, idents):
         if msg_id in self.aborted:
             # is it safe to assume a msg_id will not be resubmitted?
             self.aborted.remove(msg_id)
-            self._send_abort_reply(stream, msg, idents)
+            await self._send_abort_reply(stream, msg, idents)
             return False
         return True
 
-    async def dispatch_shell(self, msg):
-        """dispatch shell requests"""
-        if not self.session:
+    def pre_handler_hook(self):
+        """Hook to execute before calling message handler"""
+        # ensure default_int_handler during handler call
+
+    def post_handler_hook(self):
+        """Hook to execute after calling message handler"""
+
+    async def enter_eventloop(self):
+        """enter eventloop"""
+        self.log.info("Entering eventloop %s", self.eventloop)
+        # record handle, so we can check when this changes
+        eventloop = self.eventloop
+        if eventloop is None:
+            self.log.info("Exiting as there is no eventloop")
             return
-        # flush control queue before handling shell requests
-        await self._flush_control_queue()
 
-        idents, msg = self.session.feed_identities(msg, copy=False)
+        async def advance_eventloop():
+            # check if eventloop changed:
+            if self.eventloop is not eventloop:
+                self.log.info("exiting eventloop %s", eventloop)
+                return
+            self.log.debug("Advancing eventloop %s", eventloop)
+            try:
+                eventloop(self)
+            except KeyboardInterrupt:
+                # Ctrl-C shouldn't crash the kernel
+                self.log.error("KeyboardInterrupt caught in kernel")
+            if self.eventloop is eventloop:
+                # schedule advance again
+                await schedule_next()
+
+        async def schedule_next():
+            """Schedule the next advance of the eventloop"""
+            # flush the eventloop every so often,
+            # giving us a chance to handle messages in the meantime
+            self.log.debug("Scheduling eventloop advance")
+            await sleep(0.001)
+            await advance_eventloop()
```

Comment:
Since I think most of the eventloops rely on returning from the `eventloop()` call to hand control back to the kernel, it may be the case that we enter and exit from the eventloop pretty often, nesting the calls of `schedule_next` and `advance_eventloop` pretty deeply. Would that kind of potentially deep recursion be ok? It seems that the maximum depth allowed by Python is very high, so this is maybe more of an architecture question...

Reply:
Yes, we could schedule `advance_eventloop` to be called in a task instead of `await advance_eventloop()`.